README.md
# v-visibility-change

Page Visibility API wrapper for vuejs.

## Features

- no denpendencies
- support vue directive && global callback
- detect page visibility state
- full tests in chrome, theoretically compatible with IE > 10,	Firefox > 10, Safari > 6.1

## Usage

### install

#### npm

```shell
$ npm i vue-visibility-change -S
```

#### script

可以直接使用`script`标签引入
```html
<script src="./lib/vue-visibility.js"></script>
```

### global

```js
import Vue from 'vue';
import visibility from 'vue-visibility-change';

// registry directive
Vue.use(visibility);

// global mode
const handler = visibility.change((evt, hidden) => {
  // do something
});

visibility.hidden(); // Return true if page now isn’t visible to user.

visibility.unbind(handler); // Remove `change` listener by it's handler.

visibility.isSupported(); // Return true if browser support Page Visibility API.
```

### vue-directive

```html
<template>
  <div v-visibility-change="visibilityChange">
  </div>
</template>
<script>
export default {
  methods: {
    visibilityChange(hidden) {
      // do something
      console.log(hidden);
    }
  }
};
</script>
```

## Demo

visit [Demo](https://codesandbox.io/s/llok18zjy7) page, open `console pane` in the lower right corner, switch broswer tab to see output.

![demo.gif](https://i.loli.net/2018/01/09/5a53c65c73db9.gif
)

## License

[MIT](./LICENSE)

Copyright (c) 2017-present, xlaoyu

HISTORY
v1.1.0 / 2018-01-14
==================

  - feat: 添加测试，隐藏私有变量

test/index.js

// require all test files (files that ends with .spec.js)
const testsContext = require.context('./specs', true, /\.spec$/)
testsContext.keys().forEach(testsContext)

// require all src files except main.js for coverage.
// you can also change this to match only the subset of files that
// you want coverage for.
const srcContext = require.context('../src', true, /^\.\/(?!main(\.js)?$)/)
srcContext.keys().forEach(srcContext)


test/utils.js
import Vue from 'vue';

Vue.config.productionTip = false;

let id = 0;

function createElm() {
  const elm = document.createElement('div');

  elm.id = 'app' + ++id;
  document.body.appendChild(elm);

  return elm;
}

/**
 * 回收 vm
 * @param  {Object} vm
 */
function destroyVM(vm) {
  vm.$destroy && vm.$destroy();
  vm.$el &&
  vm.$el.parentNode &&
  vm.$el.parentNode.removeChild(vm.$el);
}

/**
 * 创建一个 Vue 的实例对象
 * @param  {Object|String}  Compo   组件配置，可直接传 template
 * @param  {Boolean=false} mounted 是否添加到 DOM 上
 * @return {Object} vm
 */
function createVue(Compo, mounted = false) {
  if (Object.prototype.toString.call(Compo) === '[object String]') {
    Compo = {
      template: Compo
    };
  }
  return new Vue(Compo).$mount(mounted === false ? null : createElm());
}


/**
 * 触发一个事件
 * mouseenter, mouseleave, mouseover, keyup, change, click 等
 * @param  {Element} elm
 * @param  {String} name
 * @param  {*} opts
 */
function triggerEvent(elm, name, ...opts) {
  let eventName;

  if (/^mouse|click/.test(name)) {
    eventName = 'MouseEvents';
  } else if (/^key/.test(name)) {
    eventName = 'KeyboardEvent';
  } else {
    eventName = 'HTMLEvents';
  }
  const evt = document.createEvent(eventName);

  evt.initEvent(name, ...opts);
  elm.dispatchEvent ?
    elm.dispatchEvent(evt) :
    elm.fireEvent('on' + name, evt);

  return elm;
}

/**
 * 触发 “mouseup” 和 “mousedown” 事件
 * @param {Element} elm
 * @param {*} opts
 */
function triggerClick(elm, ...opts) {
  triggerEvent(elm, 'mousedown', ...opts);
  triggerEvent(elm, 'mouseup', ...opts);

  return elm;
}

function visibilitychange() {
  // 模拟触发事件
  const event = new Event('visibilitychange');
  document.dispatchEvent(event);
}

export {
  destroyVM,
  createVue,
  triggerEvent,
  triggerClick,
  visibilitychange
};

test/karma.conf.js
// Karma configuration
// Generated on Thu Jan 11 2018 21:53:03 GMT+0800 (CST)
var webpackConfig = require('../build/webpack.test.conf');

module.exports = function (config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',

    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['Chrome'],

    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['mocha', 'power-assert'],

    // list of files / patterns to load in the browser
    files: [
      './index.js'
    ],

    // list of files / patterns to exclude
    exclude: [],

    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      './index.js': ['webpack', 'sourcemap'],
    },
    webpack: webpackConfig,
    webpackMiddleware: {
      noInfo: false
    },
    coverageReporter: {
      dir: './coverage',
      reporters: [
        { type: 'lcov', subdir: '.' },
        { type: 'text-summary' }
      ]
    },

    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['spec', 'coverage'],

    // web server port
    port: 9876,

    // enable / disable colors in the output (reporters and logs)
    colors: true,

    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,

    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,

    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false,

    // Concurrency level
    // how many browser should be started simultaneous
    concurrency: Infinity
  })
}

src/index.js
const _id = Symbol('visibility-change-id');

let hidden, visibilityChange, lastId = -1;
/* istanbul ignore else */
if (typeof document.hidden !== "undefined") { // Opera 12.10 and Firefox 18 and later support
  hidden = "hidden";
  visibilityChange = "visibilitychange";
} else if (typeof document.msHidden !== "undefined") {
  hidden = "msHidden";
  visibilityChange = "msvisibilitychange";
} else if (typeof document.webkitHidden !== "undefined") {
  hidden = "webkitHidden";
  visibilityChange = "webkitvisibilitychange";
} else if (typeof document.mozHidden !== "undefined") {
  hidden = "mozHidden";
  visibilityChange = "mozvisibilitychange";
}

// 缓存 visibility change 触发时的回调函数
const _callbacks = {};
const _doc = document;
let _init;

/**
 * Listener for `visibilitychange` event.
 * @param {Event} event event object
 */
function _change(event) {
  for ( var i in _callbacks ) {
    _callbacks[i].call(_doc, event, _doc[hidden]);
  }
}

/**
 * Set listener for `visibilitychange` event.
 */
function _listen() {
  /* istanbul ignore if */
  if (_init) {
    return;
  }

  var listener = function () {
    _change.apply(visibility, arguments);
  };
  document.addEventListener(visibilityChange, listener);

  _init = true;
}

const visibility = {

  /**
   * 当可见性发生变化时调用回调。
   * 事件监听会在第一个回调绑定时才注册，这是一个延时行为
   * 
   * @param {Function (event, hidden)} callback 回调函数
   *  - {Event} event 原始事件对象
   *  - {Boolean} hidden 表示当前页面可见性，true 表示可见
   */
  change(callback) {
    if (!visibility.isSupported()) {
        return false;
    }
    lastId += 1;
    var number = lastId;
    _callbacks[number] = callback;
    _listen();
    return number;
  },

  /**
   * Remove `change` listener by it ID.
   * 
   * @param {Number} id 回调标识
   * 
   * @example
   * var id = visibility.change(function(e, state) {
   *     firstChangeCallback();
   *     visibility.unbind(id);
   * });
   */
  unbind(id) {
    delete _callbacks[id];
  },

  /**
   * Return true if browser support Page Visibility API.
   */
  isSupported() {
    return hidden !== undefined;
  },

  /**
   * Return true if page now isn’t visible to user.
   */
  hidden() {
    return _doc[hidden];
  }
};

visibility.install = function(Vue) {

  Vue.directive('visibility-change', {
    bind(el, { value }) {
      if (typeof value === 'function') {
        el[_id] = visibility.change((evt, hidden) => {
          value(evt, hidden);
        });
      } else {
        console.error('You need bind a callback function');
      }
    },
    unbind(el) {
      visibility.unbind(el[_id]);
    }
  });
}

export default visibility;

lib/vue-visibility.js
(function webpackUniversalModuleDefinition(root, factory) {
	if(typeof exports === 'object' && typeof module === 'object')
		module.exports = factory();
	else if(typeof define === 'function' && define.amd)
		define([], factory);
	else if(typeof exports === 'object')
		exports["visibility"] = factory();
	else
		root["visibility"] = factory();
})(typeof self !== 'undefined' ? self : this, function() {
return /******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, {
/******/ 				configurable: false,
/******/ 				enumerable: true,
/******/ 				get: getter
/******/ 			});
/******/ 		}
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "";
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = 0);
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ (function(module, exports, __webpack_require__) {

"use strict";


Object.defineProperty(exports, "__esModule", {
  value: true
});
var _id = Symbol('visibility-change-id');

var _hidden = void 0,
    visibilityChange = void 0,
    lastId = -1;
/* istanbul ignore else */
if (typeof document.hidden !== "undefined") {
  // Opera 12.10 and Firefox 18 and later support
  _hidden = "hidden";
  visibilityChange = "visibilitychange";
} else if (typeof document.msHidden !== "undefined") {
  _hidden = "msHidden";
  visibilityChange = "msvisibilitychange";
} else if (typeof document.webkitHidden !== "undefined") {
  _hidden = "webkitHidden";
  visibilityChange = "webkitvisibilitychange";
} else if (typeof document.mozHidden !== "undefined") {
  _hidden = "mozHidden";
  visibilityChange = "mozvisibilitychange";
}

// 缓存 visibility change 触发时的回调函数
var _callbacks = {};
var _doc = document;
var _init = void 0;

/**
 * Listener for `visibilitychange` event.
 * @param {Event} event event object
 */
function _change(event) {
  for (var i in _callbacks) {
    _callbacks[i].call(_doc, event, _doc[_hidden]);
  }
}

/**
 * Set listener for `visibilitychange` event.
 */
function _listen() {
  /* istanbul ignore if */
  if (_init) {
    return;
  }

  var listener = function listener() {
    _change.apply(visibility, arguments);
  };
  document.addEventListener(visibilityChange, listener);

  _init = true;
}

var visibility = {

  /**
   * 当可见性发生变化时调用回调。
   * 事件监听会在第一个回调绑定时才注册，这是一个延时行为
   * 
   * @param {Function (event, hidden)} callback 回调函数
   *  - {Event} event 原始事件对象
   *  - {Boolean} hidden 表示当前页面可见性，true 表示可见
   */
  change: function change(callback) {
    if (!visibility.isSupported()) {
      return false;
    }
    lastId += 1;
    var number = lastId;
    _callbacks[number] = callback;
    _listen();
    return number;
  },


  /**
   * Remove `change` listener by it ID.
   * 
   * @param {Number} id 回调标识
   * 
   * @example
   * var id = visibility.change(function(e, state) {
   *     firstChangeCallback();
   *     visibility.unbind(id);
   * });
   */
  unbind: function unbind(id) {
    delete _callbacks[id];
  },


  /**
   * Return true if browser support Page Visibility API.
   */
  isSupported: function isSupported() {
    return _hidden !== undefined;
  },


  /**
   * Return true if page now isn’t visible to user.
   */
  hidden: function hidden() {
    return _doc[_hidden];
  }
};

visibility.install = function (Vue) {

  Vue.directive('visibility-change', {
    bind: function bind(el, _ref) {
      var value = _ref.value;

      if (typeof value === 'function') {
        el[_id] = visibility.change(function (evt, hidden) {
          value(evt, hidden);
        });
      } else {
        console.error('You need bind a callback function');
      }
    },
    unbind: function unbind(el) {
      visibility.unbind(el[_id]);
    }
  });
};

exports.default = visibility;

/***/ })
/******/ ]);
});